control-plane/api-gateway/controllers/reference_validator.go

```diff
@@ -0,0 +1,174 @@
+package controllers
+
+import (
+	"context"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gwv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+	gwv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+type ReferenceValidator struct {
+	client.Client
+}
+
+func NewReferenceValidator(client client.Client) *ReferenceValidator {
+	return &ReferenceValidator{
+		client,
+	}
+}
+
+func (rv ReferenceValidator) GatewayCanReferenceSecret(ctx context.Context, gateway gwv1beta1.Gateway, secretRef gwv1beta1.SecretObjectReference) (bool, error) {
+	fromNS := gateway.GetNamespace()
+	fromGK := metav1.GroupKind{
+		Group: gateway.GroupVersionKind().Group,
+		Kind:  gateway.GroupVersionKind().Kind,
+	}
+
+	// Kind should default to Secret if not set
+	// https://github.com/kubernetes-sigs/gateway-api/blob/v0.6.2/apis/v1beta1/object_reference_types.go#LL59C21-L59C21
+	toNS, toGK := createValuesFromRef(secretRef.Namespace, secretRef.Group, secretRef.Kind, "Secret")
+
+	return referenceAllowed(ctx, fromGK, fromNS, toGK, toNS, string(secretRef.Name), rv.Client)
+}
+
+func (rv ReferenceValidator) HTTPRouteCanReferenceGateway(ctx context.Context, httproute gwv1beta1.HTTPRoute, parentRef gwv1beta1.ParentReference) (bool, error) {
+	fromNS := httproute.GetNamespace()
+	fromGK := metav1.GroupKind{
+		Group: httproute.GroupVersionKind().Group,
+		Kind:  httproute.GroupVersionKind().Kind,
+	}
+
+	// Kind should default to Gateway if not set
+	// https://github.com/kubernetes-sigs/gateway-api/blob/v0.6.2/apis/v1beta1/shared_types.go#L48
```

Comment:
Love having these references to the spec in comments. #SheBroughtRecipts

Reply:
I cannot take the credit here, it was done in consul-api-gateway. But still a good idea whoever did it at that point!